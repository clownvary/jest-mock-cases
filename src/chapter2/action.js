import axios from 'axios';
import {
  GET_WEATHER_REQUEST,
  GET_WEATHER_FAIL,
  GET_WEATHER_SUCCESS
} from './const';

export const fetchWeatherAsyncAction = () => (dispatch) => {
  axios.get('/api/weather/city')
    .then((res) => {
      const { data } = res;
      dispatch({ type: GET_WEATHER_SUCCESS, payload: data });
    });
};

export default fetchWeatherAsyncAction;
