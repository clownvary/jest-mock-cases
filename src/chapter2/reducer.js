import {
  GET_WEATHER_REQUEST,
  GET_WEATHER_FAIL,
  GET_WEATHER_SUCCESS
} from './const';

const initialState = {
  weatherData: null,
  errorInfo: null,
  loading: false
};

function weather(state = initialState, action) {
  switch (action.type) {
    case GET_WEATHER_REQUEST:
      return {
        ...state,
        loading: true
      };
    case GET_WEATHER_SUCCESS:
      return {
        ...state,
        weatherData: action.payload
      };
    case GET_WEATHER_FAIL:
      return {
        ...state,
        errorInfo: action.payload
      };
    default:
      return state;
  }
}

export default weather;
