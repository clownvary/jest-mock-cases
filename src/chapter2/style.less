.weather {
    width: 50%;
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: auto;

    &__panel {
        width: 80%;
        padding: 10px;
        border: dashed;
        min-height: 200px;
        .error {
          color: red;
        }
    }
}