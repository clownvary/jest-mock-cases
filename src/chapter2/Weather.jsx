/* eslint-disable react/jsx-one-expression-per-line */
/* eslint-disable react/prop-types */
import React from 'react';
import { connect } from 'react-redux';
import { fetchWeatherAsyncAction } from './action';
import './style.less';

export class Weather extends React.PureComponent {

  handleClick = () => {
    this.props.fetchWeatherAsyncAction();
  }

  render() {
    const { weather: { weatherData } } = this.props;

    return (
      <div className="weather">
        <div className="weather__panel">
          {
            weatherData && (
              <div>
                <p>城市:{weatherData.city}</p>
                <p>温度:{weatherData.temperature}</p>
                <p>最高温度:{weatherData.high}</p>
                <p>最低温度:{weatherData.low}</p>
                <p>空气质量:{weatherData.quality}</p>
                <p>建议:{weatherData.advise}</p>
              </div>
            )
          }
        </div>
        <div className="weather__button">
          <button type="button" onClick={this.handleClick}>西安天气</button>
        </div>
      </div>
    );
  }
}

export default connect(
  (state) => ({
    weather: state.weather
  }),
  {
    fetchWeatherAsyncAction
  }
)(Weather);
