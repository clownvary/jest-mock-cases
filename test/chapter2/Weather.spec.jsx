/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Weather } from 'chapter2/Weather';

const initialState = {
  weatherData: null,
  errorInfo: null
};

function setup(_state = initialState) {
  const actions = {
    fetchWeatherAsyncAction: jest.fn()
  };

  const state = { ...initialState, ..._state };
  const component = shallow(<Weather weather={state} {...actions} />);

  return {
    component,
    panel: component.find('.weather__panel'),
    button: component.find('.weather__button'),
    actions
  };
}


describe('Weather component:', () => {
  it('component should render correctly', () => {
    const { component } = setup();
    expect(toJson(component)).toMatchSnapshot();
  });
  it('component should render "info" section correctly when weatherData is not null', () => {
    const mockData = {
      status: 200,
      city: '西安市',
      quality: '良',
      temperature: '20',
      advice: '极少数敏感人群应减少户外活动'
    };
    const { panel } = setup({ weatherData: mockData });
    const info = panel.find('.info');

    expect(info.length).toBe(1);
    expect(info.find('p').at(0).text()).toBe(`城市:${mockData.city}`);
    expect(info.find('p').at(1).text()).toBe(`温度:${mockData.temperature}`);
    expect(info.find('p').at(2).text()).toBe(`空气质量:${mockData.quality}`);
    expect(info.find('p').at(3).text()).toBe(`建议:${mockData.advice}`);
  });
  it('component should render "errorInfo" section correctly when errorInfo is not null', () => {
    const errorInfo = 'test error message';
    const { panel } = setup({ errorInfo });
    const error = panel.find('.error');
    expect(error.length).toBe(1);
    expect(error.text()).toBe(errorInfo);
  });
  // instance way to test
  it.skip('fetchWeatherAsyncAction should be triggered after clicked button', () => {
    const { component, actions } = setup();
    const ins = component.instance();
    ins.handleClick();
    expect(actions.fetchWeatherAsyncAction).toHaveBeenCalled();
  });
  // simulate evnet way to test
  it('fetchWeatherAsyncAction should be triggered after clicked button', () => {
    const { component, actions } = setup();
    const button = component.find('button');
    button.simulate('click');
    expect(actions.fetchWeatherAsyncAction).toHaveBeenCalled();
  });
});
