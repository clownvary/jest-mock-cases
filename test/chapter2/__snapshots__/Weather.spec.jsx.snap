// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather component: component should render correctly 1`] = `
<div
  className="weather"
>
  <div
    className="weather__panel"
  />
  <div
    className="weather__button"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      西安天气
    </button>
  </div>
</div>
`;
